### PR TITLE
Update VirtualBox Group parsing to align with documentation.

### LINKED ISSUE #8508: Virtualbox Dynamic Inventory Nested Group Assignments
### Summary

My usage of Ansible takes advantage of Ansible's nested groups for variable management. I make heavy use of Vagrant + VirtualBox to speed up local development and testing of Ansible code.

The way the dynamic inventory assigns groups currently does not match the VirtualBox documentation. According to the documentation, you can assign a VM with multiple groups by delimiting it with a comma `,` character, and nest groups using the slash `/` character. 

Ultimately, the objective is to be able to harmonize the way groups are set up with VirtualBox and with Ansible. It appears the concept of nested (or parent-child) groups match, so it should align.

See: https://www.virtualbox.org/manual/UserManual.html#gui-vmgroups and https://docs.ansible.com/ansible/latest/inventory_guide/intro_inventory.html#how-variables-are-merged

From VirtualBox Docs:

```shell
# Create multiple groups. For example:
VBoxManage modifyvm "vm01" --groups "/TestGroup,/TestGroup2"

#Create nested groups, having a group hierarchy. For example:
VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2"

```

The way the dynamic inventory parses the groups considers the `/` character as the delimiter for group assignment. As a consequence, we are unable to use nested groups in Ansible. If a user does not care about nested group, and assign multiple groups by simply using the `/` character (e.g. `/TestGroup/TestGroup2`), the VirtualBox UI will show a deeply nested list of groups before displaying the actual VM. See vm-1 in the screenshot for an example.

Screenshot_20240614_174813

The example I provide shows 2 issues resulting from how groups are parsed:

vm-1 is not set up with nested groups, but is set up to participate in 3 top-level groups.
vm-2 groups have a trailing comma (except for the last one).

I did not spot mentions about how groups are automatically inferred in the official docs: https://docs.ansible.com/ansible/latest/collections/community/general/virtualbox_inventory.html

### Issue Type

Bug Report

### Component Name

virtualbox inventory

### Ansible Version

```console
ansible [core 2.17.0]
  config file = None
  configured module search path = ['/home/rly/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/rly/Documents/dev/sandbox/.venv/lib/python3.12/site-packages/ansible
  ansible collection location = /home/rly/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/rly/Documents/dev/sandbox/.venv/bin/ansible
  python version = 3.12.3 (main, Apr 23 2024, 09:16:07) [GCC 13.2.1 20240417] (/home/rly/Documents/dev/sandbox/.venv/bin/python)
  jinja version = 3.1.4
  libyaml = True
```

### Community.general Version

```console
# /home/rly/Documents/dev/sandbox/.venv/lib/python3.12/site-packages/ansible_collections
Collection        Version
----------------- -------
community.general 9.0.1
```

### Configuration

```console
CONFIG_FILE() = None
```

### OS / Environment

My Host OS: 6.9.1-arch1-1
Vagrant: 2.4.1
VirtualBox: 7.0.18 r162988
VirtualBox VMs: bento/rockylinux-9 Vagrant images

### Steps to Reproduce

I made a sample `Vagrantfile` to help illustrate. Notice that I'm purposefully trying to make the hierarchy of groups for vm-1 to be the reverse from lexicographical ordering.

Ensure your VirtualBox host has a host-only network named vboxnet0 with `192.168.56.1/24` set up. The `Vagrantfile` will also copy your SSH key into the VMs for convenience when using with Ansible.

```ruby
Vagrant.configure("2") do |config|
  DEFAULT_VAGRANT_BOX = "bento/rockylinux-9"
  
  config.vm.define "vm-1" do |config|
    hostname = "vm-1"
    config.vm.box = DEFAULT_VAGRANT_BOX
    config.vm.hostname = hostname
    config.vm.network "private_network", ip: "192.168.56.10", name: "vboxnet0"

    config.vm.provider "virtualbox" do |vb|
      vb.name = hostname
      vb.customize ["modifyvm", :id, "--groups", "/zgroup1/ygroup2/xgroup3"]
    end

    config.vm.provision "shell", run: "always" do |s|
      ssh_pub_key = File.readlines("#{Dir.home}/.ssh/id_rsa.pub").first.strip
      s.inline = <<-SHELL
        echo "SSH Public key will be added to /home/vagrant/.ssh/authorized_keys"
        echo '#{ssh_pub_key}' >> /home/vagrant/.ssh/authorized_keys
      SHELL
    end
  end

  config.vm.define "vm-2" do |config|
    hostname = "vm-2"
    config.vm.box = DEFAULT_VAGRANT_BOX
    config.vm.hostname = hostname
    config.vm.network "private_network", ip: "192.168.56.11", name: "vboxnet0"
    config.vm.provider "virtualbox" do |vb|
      vb.name = hostname
      vb.customize ["modifyvm", :id, "--groups", "/othergroup1,/othergroup2,/othergroup3"]
    end

    config.vm.provision "shell", run: "always" do |s|
      ssh_pub_key = File.readlines("#{Dir.home}/.ssh/id_rsa.pub").first.strip
      s.inline = <<-SHELL
        echo "SSH Public key will be added to /home/vagrant/.ssh/authorized_keys"
        echo '#{ssh_pub_key}' >> /home/vagrant/.ssh/authorized_keys
      SHELL
    end
  end
end
```

And my simple inventory file
```yaml
# virtualbox.yml
plugin: community.general.virtualbox

network_info_path: /VirtualBox/GuestInfo/Net/1/V4/IP 
running_only: true

compose:
  ansible_user: "'vagrant'"
```

And some sample group vars
```console
grep -R my_custom_variable group_vars/ | sort

group_vars/all.yml:my_custom_variable: This is in "all" group
group_vars/xgroup3.yml:my_custom_variable: This is in "xgroup3" group
group_vars/ygroup2.yml:my_custom_variable: This is in "ygroup2" group
group_vars/zgroup1.yml:my_custom_variable: This is in "zgroup1" group
```

```console
vagrant up
ansible -i virtualbox.yml -m debug -a "var=group" all
ansible -i virtualbox.yml -m debug -a "var=group_names" all
ansible -i virtualbox.yml -m debug -a "var=my_custom_variable" all
```

### Expected Results

I expected to see vm-1 have its variable resolved properly, in consideration for nested groups when using `/zgroup1/ygroup2/xgroup3`, and for vm-2 to be part of separate groups when using `/othergroup1,/othergroup2,/othergroup3`

For vm-1, if we were to consider this in a static inventory, it might look like this:
```yaml
all:
  vars:
    my_custom_variable: This is in "all" group
  children:
    zgroup1:
      vars:
        my_custom_variable: This is in "zgroup1" group
      children:
        ygroup2:
          vars:
            my_custom_variable: This is in "ygroup2" group
          children:
            xgroup3:
              vars:
                my_custom_variable: This is in "xgroup3" group
              hosts:
                vm-1:
                  ansible_host: 192.168.56.10
                  ansible_user: vagrant
```

and 

```console
ansible -i static.yml -m debug -a "var=my_custom_variable" all
vm-1 | SUCCESS => {
    "my_custom_variable": "This is in \"xgroup3\" group"
}
```

The same would be true with the dynamic inventory.

```console
ansible -i virtualbox.yml -m debug -a "var=my_custom_variable" vm-1
vm-1 | SUCCESS => {
    "my_custom_variable": "This is in \"xgroup3\" group"
}
```

For vm-2, it should be part of the `othergroup1`, `othergroup2`, `othergroup3` groups.

```console
ansible -i virtualbox.yml -m debug -a "var=group_names" vm-2
vm-2 | SUCCESS => {
    "group_names": [
        "othergroup1",
        "othergroup2",
        "othergroup3"
    ]
}
```

### Actual Results

```console
ansible -i virtualbox.yml -m debug -a "var=group_names" all

vm-1 | SUCCESS => {
    "group_names": [
        "xgroup3",
        "ygroup2",
        "zgroup1"
    ]
}
vm-2 | SUCCESS => {
    "group_names": [
        "othergroup1,",
        "othergroup2,",
        "othergroup3"
    ]
}
```

In this case, vm-1 appears to be correctly assigned, but watch out!

```console
ansible -i virtualbox.yml -m debug -a "var=my_custom_variable" all
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details
vm-1 | SUCCESS => {
    "my_custom_variable": "This is in \"zgroup1\" group"
}
vm-2 | SUCCESS => {
    "my_custom_variable": "This is in \"all\" group"
}
```
The `zgroup1` group vars wins because it's ordered last, lexicographically. This shows they're considered to be "on the same level".

Also, vm-2 groups have a trailing `,` character which is incorrect. This is due to the code simply splitting on the `/` character, so the `,` remains.

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by lyrandy**
Previously, we could separate the group string on the `/` char and consider each element to be distinct, top-level groups. This change implements the notion of nested groups and the use of the `,` char to split multiple groups.

##### SUMMARY
Fixes #8508 
Update the parsing algorithm in the VirtualBox dynamic inventory. The update changes how groups are formed when considering VirtualBox VMs.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
virtualbox inventory plugin

##### ADDITIONAL INFORMATION
Before, we would split groups based only on the `/` characters to form top-level groups. This was problematic for users (like me) who want to use nested groups for better variable management.

This change introduces the ability to use nested groups using `/`, and different groups using `,`. This approach more closely aligns with the official VirtualBox documentation.

https://www.virtualbox.org/manual/UserManual.html#gui-vmgroups

In the linked issue, it shows that previously, the `my_custom_variable` would resolve to `"This is in \"zgroup1\" group"` because the groups for vm-1 are on "the same level", and so variable resolution depends on lexicographical ordering.
Also, groups for vm-2 had a trailing `,` character.

Here are the results of the Ansible debug commands with this updated inventory file.

```console
ansible -i virtualbox.yml -m debug -a "var=my_custom_variable" all
vm-1 | SUCCESS => {
    "my_custom_variable": "This is in \"xgroup3\" group"
}
vm-2 | SUCCESS => {
    "my_custom_variable": "This is in \"all\" group"
}

ansible -i virtualbox.yml -m debug -a "var=group_names" all
vm-1 | SUCCESS => {
    "group_names": [
        "xgroup3",
        "ygroup2",
        "zgroup1"
    ]
}
vm-2 | SUCCESS => {
    "group_names": [
        "othergroup1",
        "othergroup2",
        "othergroup3"
    ]
}
```

### CODE REVIEW

**Comment by felixfontein** [comment]
Thanks for your contribution!

I'm pretty sure this is a new feature, and not a bugfix, since the documentation of the inventory plugin never mentioned the new behavior. It is a new feature, and (as you correctly stated in the changelog fragment) a breaking change. Breaking changes are not acceptable without a long enough deprecation period informing users of the changing behavior.

The best choice is to make the behavior configurable, with default being the old behavior. Then users who want can switch to the new behavior, and it isn't breaking anyone.

File: changelogs/fragments/8508-virtualbox-inventory.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+minor_changes:
+  - >- 
+    virtualbox inventory plugin - expose a new parameter `enable_advanced_group_parsing` to change how the VirtualBox dynamic inventory parses VM groups.
+    The dynamic inventory automatically assigns VMs to Ansible Groups based on VirtualBox groups configured using VBoxManage.
+    When set to `false` (default, existing behaviour), the `/` character would delineate the different groups.
+    When set to `true`, the `/` character allows creation of nested groups, and the `,` character is used to delineate the groups.
+    See https://www.virtualbox.org/manual/UserManual.html#gui-vmgroups for details on how to assign VirtualBox groups. (https://github.com/ansible-collections/community.general/issues/8508)
```
Comment (felixfontein) [inline changelogs/fragments/8508-virtualbox-inventory.yml:7]: I think it's OK to shorten this, users can look up the documentation of the plugin to see how the new option works.
```suggestion
    virtualbox inventory plugin - expose a new parameter ``enable_advanced_group_parsing`` to change how the VirtualBox dynamic inventory parses VM groups (https://github.com/ansible-collections/community.general/issues/8508).
```

File: plugins/inventory/virtualbox.py [hunk 2]
```diff
@@ -35,6 +37,17 @@
...
+        enable_advanced_group_parsing:
+            description: >
+                The default group parsing rule (when this setting is set to `false`) is to split the VirtualBox VM's group based on the `/` character and
+                assign the resulting list elements as an Ansible Group.
+                Setting `enable_advanced_group_parsing` to `true` changes this behaviour to match VirtualBox's interpretation of groups according to:
+                https://www.virtualbox.org/manual/UserManual.html#gui-vmgroups
+                Groups are now split using the `,` character, and the `/` character indicates nested groups.
+                So, for a VM that's been configured using `VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2,/TestGroup3"`:
+                - TestGroup2 is a child group of TestGroup
+                - The VM will be part of TestGroup2 and TestGroup3.
```
Comment (felixfontein) [inline plugins/inventory/virtualbox.py:49]: This kind of formatting won't work.
```suggestion
            description:
              - The default group parsing rule (when this setting is set to V(false)) is to split the VirtualBox VM's group based on the V(/) character and
                assign the resulting list elements as an Ansible Group.
              - Setting O(enable_advanced_group_parsing=true) changes this behaviour to match VirtualBox's interpretation of groups according to
                U(https://www.virtualbox.org/manual/UserManual.html#gui-vmgroups).
                Groups are now split using the V(,) character, and the V(/) character indicates nested groups.
              - So, for a VM that's been configured using V(VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2,/TestGroup3"):
                C(TestGroup2) is a child group of C(TestGroup);
                the VM will be part of C(TestGroup2) and C(TestGroup3).
```
(Also check out https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#semantic-markup-within-module-documentation)

File: plugins/inventory/virtualbox.py [hunk 3]
```diff
@@ -35,6 +37,17 @@
...
+            default: false
```
Comment (felixfontein) [inline plugins/inventory/virtualbox.py:50]:
```suggestion
            default: false
            type: bool
```

File: plugins/inventory/virtualbox.py [hunk 4]
```diff
@@ -227,6 +236,63 @@ def find_host(host, inventory):
...
+                # possible assignments from VirtualBox. e.g. ,/Group1
+                continue
+
+            if group == "/":
+                # This is the "root" group. We get here if the VM was not
+                # assigned to a particular group. Consider the host to be
+                # unassigned to a group.
+                continue
+
+            parent_group = "all"
```
Comment (felixfontein) [inline plugins/inventory/virtualbox.py:274]: I would start with `None` here and not add the group explicitly to the `all` group.
```suggestion
            parent_group = None
```

File: plugins/inventory/virtualbox.py [hunk 5]
```diff
@@ -227,6 +236,63 @@ def find_host(host, inventory):
...
+                    continue
+
+                if subgroup == '/':
+                    # "root" group.
+                    # Consider the host to be unassigned
+                    continue
+
+                subgroup = make_unsafe(subgroup)
+                subgroup = self.inventory.add_group(subgroup)
+                self.inventory.add_child(parent_group, subgroup)
```
Comment (felixfontein) [inline plugins/inventory/virtualbox.py:288]:
```suggestion
                if parent_group is not None:
                    self.inventory.add_child(parent_group, subgroup)
```

File: plugins/inventory/virtualbox.py (hunk 2 above)
Comment (lyrandy) [reply plugins/inventory/virtualbox.py:49]: Thanks for the link! It's easy to get lost in the various links to all the contributor guides. I'll bookmark this useful page!

**Comment by lyrandy** [comment]
Thank you for the great feedback Felix! I've taken in your suggestions.

**Comment by lyrandy** [comment]
The tests did not like the line ending with the colon character.
```
- So, for a VM that's been configured using V(VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2,/TestGroup3"):
  C(TestGroup2) is a child group of C(TestGroup); 
  the VM will be part of C(TestGroup2) and C(TestGroup3).
```

I will reword it to avoid that stylistic choice.

File: changelogs/fragments/8508-virtualbox-inventory.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+  - >- 
+    virtualbox inventory plugin - expose a new parameter ``enable_advanced_group_parsing`` to change how the VirtualBox dynamic inventory parses VM groups (https://github.com/ansible-collections/community.general/issues/8508).
```
Comment (felixfontein) [inline changelogs/fragments/8508-virtualbox-inventory.yml:3]:
```suggestion
    virtualbox inventory plugin - expose a new parameter ``enable_advanced_group_parsing`` to change how the VirtualBox dynamic inventory parses VM groups (https://github.com/ansible-collections/community.general/issues/8508, https://github.com/ansible-collections/community.general/pull/8510).
```

File: plugins/inventory/virtualbox.py [hunk 7]
```diff
@@ -14,6 +14,8 @@
         - Get inventory hosts from the local virtualbox installation.
         - Uses a YAML configuration file that ends with virtualbox.(yml|yaml) or vbox.(yml|yaml).
         - The inventory_hostname is always the 'Name' of the virtualbox instance.
+        - Groups can be assigned to the VMs using `VBoxManage`. Multiple groups can be assigned by using `/` as a delimeter.
+        - A separate parameter, `enable_advanced_group_parsing` is exposed to change grouping behaviour. See the parameter documentation for details.
```
Comment (felixfontein) [inline plugins/inventory/virtualbox.py:18]:
```suggestion
        - Groups can be assigned to the VMs using C(VBoxManage). Multiple groups can be assigned by using V(/) as a delimeter.
        - A separate parameter, O(enable_advanced_group_parsing) is exposed to change grouping behaviour. See the parameter documentation for details.
```

**Comment by felixfontein** [comment]
> The tests did not like the line ending with the colon character.
> 
> ```
> - So, for a VM that's been configured using V(VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2,/TestGroup3"):
>   C(TestGroup2) is a child group of C(TestGroup); 
>   the VM will be part of C(TestGroup2) and C(TestGroup3).
> ```

That's basic YAML parsing which doesn't like it. A colon followed by a space indicates a dictionary entry in YAML, so you end up with a one-element dictionary with a funny key and value instead of having a string. Simply quoting the thing (or using things like `>-` etc.) prevents that.

File: plugins/inventory/virtualbox.py (hunk 7 above)
Comment (lyrandy) [reply plugins/inventory/virtualbox.py:18]: I'm embarrassed I missed this part even though you were kind enough to show me the docs last time. I'll have to be more diligent in the future.

**Comment by lyrandy** [comment]
> > The tests did not like the line ending with the colon character.
> > ```
> > - So, for a VM that's been configured using V(VBoxManage modifyvm "vm01" --groups "/TestGroup/TestGroup2,/TestGroup3"):
> >   C(TestGroup2) is a child group of C(TestGroup); 
> >   the VM will be part of C(TestGroup2) and C(TestGroup3).
> > ```
> 
> That's basic YAML parsing which doesn't like it. A colon followed by a space indicates a dictionary entry in YAML, so you end up with a one-element dictionary with a funny key and value instead of having a string. Simply quoting the thing (or using things like `>-` etc.) prevents that.

Your explanation makes it so obvious. Thanks!

**Comment by felixfontein** [comment]
If nobody objects, I'll merge this on Sunday.

**Comment by felixfontein** [comment]
@lyrandy thanks for your contribution!

**Comment by lyrandy** [comment]
> @lyrandy thanks for your contribution!

Thank you for your guidance and help!
